src/box/tuple.h

```diff
@@ -440,18 +470,39 @@ tuple_is_unreferenced(struct tuple *tuple)
 	return tuple->local_refs == 0;
 }
 
+/** Check that the tuple is in compact mode. */
+static inline bool
+tuple_is_compact(struct tuple *tuple)
+{
+	return tuple->data_offset_bsize_raw & 0x8000;
+}
+
 /** Offset to the MessagePack from the beginning of the tuple. */
 static inline uint16_t
 tuple_data_offset(struct tuple *tuple)
 {
-	return tuple->data_offset;
+	uint16_t res = tuple->data_offset_bsize_raw;
+	res = (res & 0x7fff) >> ((res >> 15) * 8);
+#ifndef NDEBUG
+	uint16_t simple = (tuple->data_offset_bsize_raw & 0x8000) ?
+			  (tuple->data_offset_bsize_raw >> 8) & 0x7f :
+			  tuple->data_offset_bsize_raw;
+	assert(res == simple);
+#endif
+	return res;
 }
 
 /** Size of MessagePack data of the tuple. */
-static inline size_t
+static inline uint32_t
 tuple_bsize(struct tuple *tuple)
 {
-	return tuple->bsize;
+	if (tuple->data_offset_bsize_raw & 0x8000) {
+		return tuple->data_offset_bsize_raw & 0xff;
+	} else {
+		uint32_t bsize;
+		memcpy(&bsize, &tuple->bsize_bulky, sizeof(bsize));
```

Comment:
Why not simply

```C
bsize = tuple->bsize_bulky;
```

?

Reply:
If this code works fine without memcpy on our CI (including ARM), I think we should remove it.

Reply:
You're right, fixed.